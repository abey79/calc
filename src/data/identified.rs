//! Generic type for things that have an ID, such as AST nodes.
//!
//! This type as both a generic type for the "kind" (the content of the structure) and the ID. This
//! ensure that NodeIds are not used to look up, e.g., tokens.

use std::fmt;
use std::fmt::{Debug, Display};
use std::sync::atomic::{AtomicUsize, Ordering};

#[derive(Clone, Copy, PartialEq, Default)]
pub struct Identified<K, I> {
    pub kind: K,
    pub id: I,
}

impl<K: Display, I> Display for Identified<K, I> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        self.kind.fmt(f)
    }
}

impl<K: Debug, I: Display> Debug for Identified<K, I> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "[{}] ", self.id)?;
        self.kind.fmt(f)
    }
}

impl<K, I> AsRef<K> for Identified<K, I> {
    fn as_ref(&self) -> &K {
        &self.kind
    }
}

pub fn new_id() -> usize {
    static COUNTER: AtomicUsize = AtomicUsize::new(0);
    COUNTER.fetch_add(1, Ordering::Relaxed)
}
