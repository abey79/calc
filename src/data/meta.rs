//! Generic type for things that have an ID, such as AST nodes.
//!
//! This type as both a generic type for the "kind" (the content of the structure) and the ID. This
//! ensure that NodeIds are not used to look up, e.g., tokens.

use std::fmt;
use std::fmt::{Debug, Display};

#[derive(Clone, PartialEq, Default)]
pub struct Meta<K, I> {
    pub kind: K,
    pub meta: I,
}

impl<K: Display, I> Display for Meta<K, I> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        self.kind.fmt(f)
    }
}

impl<K: Debug, I: Display> Debug for Meta<K, I> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "[{}] ", self.meta)?;
        self.kind.fmt(f)
    }
}

impl<K, I> AsRef<K> for Meta<K, I> {
    fn as_ref(&self) -> &K {
        &self.kind
    }
}
