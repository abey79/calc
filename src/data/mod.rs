pub mod ast;
pub mod identified;
pub mod span;
pub mod token;
