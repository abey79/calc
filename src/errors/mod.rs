pub mod error_context;

use error_context::ErrorContext;
use thiserror::Error;

/// Errors generated by the tokenizer
#[derive(Error, Debug, PartialEq)]
pub enum SyntaxError {
    #[error("unexpected end of file")]
    UnexpectedEndOfFile,

    #[error("unexpected character '{0}'")]
    UnexpectedCharacter(char),
}

/// Errors generated by the tokenizer
#[derive(Error, Debug, PartialEq)]
pub enum TokenizerError {
    #[error("{1}Syntax error:{0}")]
    SyntaxError(SyntaxError, ErrorContext),
}
