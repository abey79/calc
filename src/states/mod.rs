//! This module models states (as in "state machine") between the compilation steps.
//!
//! The data associated with each state is factored in "context" structure for reusability (e.g. the
//! [`TokenStream`] context ("state data") is used in both the [`TokenizedState`] and the
//! [`ParsedState`]).

pub mod input_state;
pub mod parsed_state;
pub mod tokenized_state;

pub use input_state::*;
pub use parsed_state::*;
pub use tokenized_state::*;
