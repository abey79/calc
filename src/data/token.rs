//! Tokens

use crate::data::meta::Meta;
use crate::data::span::Span;
use std::fmt;

#[derive(Debug, Clone, PartialEq)]
pub enum TokenKind {
    Name(String),
    Int(i32),
    Float(f64),

    // misc
    Semi,
    Assign,
    LParen,
    RParen,
    Comma,

    // operators
    Plus,
    Minus,
    Star,
    Slash,

    // keywords
    Print,
}

impl fmt::Display for TokenKind {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        use TokenKind::*;
        match self {
            Name(ref s) => write!(f, "'{}'", s),
            Int(i) => write!(f, "'{}'", i),
            Float(fl) => write!(f, "'{:?}'", fl),
            Semi => write!(f, "';'"),
            Assign => write!(f, "'='"),
            LParen => write!(f, "'('"),
            RParen => write!(f, "')'"),
            Comma => write!(f, "','"),
            Plus => write!(f, "'+'"),
            Minus => write!(f, "'-'"),
            Star => write!(f, "'*'"),
            Slash => write!(f, "'/'"),
            Print => write!(f, "'print'"),
        }
    }
}

/// Token type.
///
/// This is an RC pointer to a token kind, as token will be passed around a lot.
pub type Token = Meta<TokenKind, Span>;

impl Token {
    pub fn new(kind: TokenKind, span: Span) -> Self {
        Self { kind, meta: span }
    }

    pub fn span(&self) -> Span {
        self.meta
    }
}
