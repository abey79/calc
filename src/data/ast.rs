//! This module contains the AST data structures.
//!
//! Throughout the code, the "Node" terminology is used to refer to elements of the AST. AST nodes
//! are assigned a unique `NodeId` when created. Factory functions are also provided to make it easy
//! to build AST nodes.

use crate::data::meta::Meta;
use crate::data::token::TokenKind;
use std::fmt;
use std::fmt::{Debug, Display};

// =================================================================================================
// DATA STRUCTURES

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum UnaryOpKind {
    Pos,
    Neg,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum BinOpKind {
    Add,
    Sub,
    Mul,
    Div,
}

#[derive(Debug, Clone, PartialEq)]
pub enum ExprKind<T: Debug + Display> {
    Variable(VarName<T>),
    BinOp {
        op: BinOp<T>,
        left: Box<Expr<T>>,
        right: Box<Expr<T>>,
    },
    UnaryOp {
        op: UnaryOp<T>,
        operand: Box<Expr<T>>,
    },
    Tuple(Vec<Expr<T>>),
    Integer(i32),
    Float(f64),
}

#[derive(Debug, Clone, PartialEq)]
pub enum StmtKind<T: Debug + Display> {
    Assign { name: VarName<T>, value: Expr<T> },
    Print { expr: Expr<T> },
    Expr { expr: Expr<T> },
}

pub type BinOp<T> = Meta<BinOpKind, T>;
pub type UnaryOp<T> = Meta<UnaryOpKind, T>;
pub type VarName<T> = Meta<String, T>;
pub type Expr<T> = Meta<ExprKind<T>, T>;
pub type Stmt<T> = Meta<StmtKind<T>, T>;

// =================================================================================================
// FACTORIES

impl From<&TokenKind> for UnaryOpKind {
    fn from(value: &TokenKind) -> Self {
        match value {
            TokenKind::Plus => Self::Pos,
            TokenKind::Minus => Self::Neg,
            _ => panic!("Invalid token kind: {:?}", value),
        }
    }
}

impl<T> UnaryOp<T> {
    pub fn new(kind: impl Into<UnaryOpKind>, meta: impl Into<T>) -> Self {
        Self {
            kind: kind.into(),
            meta: meta.into(),
        }
    }
}

impl From<&TokenKind> for BinOpKind {
    fn from(value: &TokenKind) -> Self {
        match value {
            TokenKind::Plus => Self::Add,
            TokenKind::Minus => Self::Sub,
            TokenKind::Star => Self::Mul,
            TokenKind::Slash => Self::Div,
            _ => panic!("Invalid token kind: {:?}", value),
        }
    }
}

impl<T> BinOp<T> {
    pub fn new(kind: impl Into<BinOpKind>, meta: impl Into<T>) -> Self {
        Self {
            kind: kind.into(),
            meta: meta.into(),
        }
    }
}

impl<T> VarName<T> {
    pub fn new(name: impl Into<String>, meta: impl Into<T>) -> Self {
        Self {
            kind: name.into(),
            meta: meta.into(),
        }
    }
}

impl<T: Debug + Display> Expr<T> {
    pub fn variable(name: impl Into<VarName<T>>, meta: impl Into<T>) -> Self {
        Self {
            kind: ExprKind::Variable(name.into()),
            meta: meta.into(),
        }
    }

    pub fn bin_op(op: BinOp<T>, left: Expr<T>, right: Expr<T>, meta: impl Into<T>) -> Self {
        Self {
            kind: ExprKind::BinOp {
                op,
                left: Box::new(left),
                right: Box::new(right),
            },
            meta: meta.into(),
        }
    }

    pub fn unary_op(op: UnaryOp<T>, operand: Expr<T>, meta: impl Into<T>) -> Self {
        Self {
            kind: ExprKind::UnaryOp {
                op,
                operand: Box::new(operand),
            },
            meta: meta.into(),
        }
    }

    pub fn tuple(exprs: Vec<Expr<T>>, meta: impl Into<T>) -> Self {
        Self {
            kind: ExprKind::Tuple(exprs),
            meta: meta.into(),
        }
    }

    pub fn integer(value: i32, meta: impl Into<T>) -> Self {
        Self {
            kind: ExprKind::Integer(value),
            meta: meta.into(),
        }
    }

    pub fn float(value: f64, meta: impl Into<T>) -> Self {
        Self {
            kind: ExprKind::Float(value),
            meta: meta.into(),
        }
    }
}

impl<T: Debug + Display> Stmt<T> {
    pub fn assign(name: impl Into<VarName<T>>, value: Expr<T>, meta: impl Into<T>) -> Self {
        Self {
            kind: StmtKind::Assign {
                name: name.into(),
                value,
            },
            meta: meta.into(),
        }
    }

    pub fn print(expr: Expr<T>, meta: impl Into<T>) -> Self {
        Self {
            kind: StmtKind::Print { expr },
            meta: meta.into(),
        }
    }

    pub fn expr(expr: Expr<T>, meta: impl Into<T>) -> Self {
        Self {
            kind: StmtKind::Expr { expr },
            meta: meta.into(),
        }
    }
}

// =================================================================================================
// UTILITY TRAITS

impl fmt::Display for BinOpKind {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self::Add => write!(f, "+"),
            Self::Sub => write!(f, "-"),
            Self::Mul => write!(f, "*"),
            Self::Div => write!(f, "/"),
        }
    }
}

impl fmt::Display for UnaryOpKind {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self::Neg => write!(f, "-"),
            Self::Pos => write!(f, "+"),
        }
    }
}
