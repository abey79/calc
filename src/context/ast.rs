use crate::data::ast::Stmt;
use crate::data::meta::Meta;
use crate::data::span::Span;
use crate::data::token_span::TokSpan;
use crate::pipeline;
use std::fmt;
use std::fmt::{Debug, Display, Write};

#[derive(Debug, Default)]
pub struct Ast<M: Debug + Display> {
    stmts: Vec<Stmt<M>>,
}

impl<M: Debug + Display> Ast<M> {
    pub fn new() -> Self {
        Self { stmts: Vec::new() }
    }

    pub fn stmts(&self) -> &[Stmt<M>] {
        &self.stmts
    }

    pub fn stmts_mut(&mut self) -> &mut Vec<Stmt<M>> {
        &mut self.stmts
    }

    pub fn push_stmt(&mut self, stmt: Stmt<M>) {
        self.stmts.push(stmt);
    }

    pub fn format<W: fmt::Write>(&self, w: &mut W) -> Result<String, fmt::Error> {
        pipeline::formatter::format(self, w)
    }
}

impl Ast<TokSpan> {
    pub fn dump<W: Write>(&self, w: &mut W) -> std::fmt::Result {
        for stmt in self.stmts() {
            writeln!(w, "\n{:#?}", stmt)?;
        }

        Ok(())
    }
}

// Provide convenient methods for AST nodes parametrized by a token span.
impl<K> Meta<K, TokSpan> {
    pub fn tok_span(&self) -> TokSpan {
        self.meta.clone()
    }

    pub fn span(&self) -> Span {
        self.meta.span()
    }
}
