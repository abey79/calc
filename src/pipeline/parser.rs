use crate::data::ast::{BinOp, Expr, Stmt, UnaryOp, VarName};
use crate::data::token::{Token, TokenId, TokenKind};
use crate::errors::{ParserError, SyntaxError};
use crate::states::{AstContext, ParsedInput, TokenizedInput};
use std::rc::Rc;

type Result<T> = std::result::Result<T, ParserError>;

pub(crate) fn parse(input: TokenizedInput) -> Result<ParsedInput> {
    let mut parser = Parser::new(input);
    parser.run()?;
    Ok(ParsedInput {
        text_ctx: parser.input.text_ctx,
        token_ctx: parser.input.token_ctx,
        ast_ctx: parser.ast_ctx,
    })
}

/// Returns the next token if it matches and advances the parser, or returns an error.
macro_rules! expect {
    ($self:ident, $kind:pat) => {{
        let token = $self
            .tokens()
            .get($self.pos)
            .ok_or_else(|| $self.end_of_file_err())?;

        if matches!(token.kind, $kind) {
            Ok($self.next().unwrap())
        } else {
            Err(ParserError::SyntaxError(
                SyntaxError::UnexpectedToken(token.kind.clone()),
                $self
                    .input
                    .text_ctx
                    .error_context($self.input.token_ctx.span_from_id(token.id)),
            ))
        }
    }};
}

/// Returns an `Option<Token>` if the token matches, in which case the parser advances
macro_rules! accept {
    ($self:ident, $kind:pat) => {{
        $self.pos += 1; // must be moved before the get to avoid double-borrowing
        let token = $self.tokens().get($self.pos - 1);
        if let Some(token) = token {
            if matches!(token.kind, $kind) {
                Some(token)
            } else {
                $self.pos -= 1;
                None
            }
        } else {
            $self.pos -= 1;
            None
        }
    }};
}

/// Macro to reduce the boilerplate associated with saving the token span of the parsed node.
macro_rules! parse_node {
    ($self:ident, $code:block) => {{
        let start = $self.cur_id();
        let node = { $code };
        $self.ast_ctx.push_span(node.id, start, $self.prev_id());
        Ok(node)
    }};
}

struct Parser {
    input: TokenizedInput,
    ast_ctx: AstContext,

    // state
    pos: usize,
}

impl Parser {
    fn new(input: TokenizedInput) -> Self {
        Self {
            input,
            ast_ctx: AstContext::default(),
            pos: 0,
        }
    }

    #[inline]
    fn tokens(&self) -> &[Rc<Token>] {
        &self.input.tokens()
    }

    fn peek(&self) -> Option<&TokenKind> {
        self.tokens().get(self.pos).map(|t| &t.kind)
    }

    fn next(&mut self) -> Option<Rc<Token>> {
        self.pos += 1;
        let token = self.tokens().get(self.pos - 1).map(|t| t.clone());
        token
    }

    fn cur_id(&self) -> Option<TokenId> {
        self.tokens().get(self.pos).map(|t| t.id)
    }

    fn prev_id(&self) -> Option<TokenId> {
        self.tokens().get(self.pos - 1).map(|t| t.id)
    }

    fn run(&mut self) -> Result<()> {
        while self.pos < self.tokens().len() {
            let start = self.cur_id();
            let stmt = self.parse_stmt()?;

            self.ast_ctx.push_span(stmt.id, start, self.prev_id());
            self.ast_ctx.push_stmt(stmt);
        }

        Ok(())
    }

    fn parse_stmt(&mut self) -> Result<Stmt> {
        let start_pos = self.pos;
        match self.peek() {
            Some(TokenKind::Print) => self.parse_print_stmt(),
            Some(TokenKind::Name(_)) => {
                // here an expr stmt could be confused with an assignment stmt
                let res = self.parse_assign_stmt();
                if res.is_ok() {
                    res
                } else {
                    self.pos = start_pos;
                    self.parse_expr_stmt()
                }
            }
            Some(_) => self.parse_expr_stmt(),
            None => Err(self.end_of_file_err()),
        }
    }

    fn parse_print_stmt(&mut self) -> Result<Stmt> {
        parse_node!(self, {
            expect!(self, TokenKind::Print)?;
            let expr = self.parse_expr()?;
            expect!(self, TokenKind::Semi)?;

            Stmt::print(expr)
        })
    }

    fn parse_assign_stmt(&mut self) -> Result<Stmt> {
        parse_node!(self, {
            let name = self.parse_var_name()?;
            expect!(self, TokenKind::Assign)?;
            let expr = self.parse_expr()?;
            expect!(self, TokenKind::Semi)?;

            Stmt::assign(name, expr)
        })
    }

    fn parse_expr_stmt(&mut self) -> Result<Stmt> {
        parse_node!(self, {
            let expr = self.parse_expr()?;
            expect!(self, TokenKind::Semi)?;

            Stmt::expr(expr)
        })
    }

    fn parse_expr(&mut self) -> Result<Expr> {
        self.parse_add_term()
    }

    fn parse_add_term(&mut self) -> Result<Expr> {
        let mut start = self.cur_id();
        let mut lhs = self.parse_mul_term()?;
        while let Some(op_token) = accept!(self, TokenKind::Plus | TokenKind::Minus) {
            // create binop node and push span
            let op = BinOp::new(&op_token.kind);
            self.ast_ctx
                .push_span(op.id, Some(op_token.id), Some(op_token.id));

            let rhs = self.parse_mul_term()?;
            lhs = Expr::bin_op(op, lhs, rhs);
            self.ast_ctx.push_span(lhs.id, start, self.prev_id());
            start = self.cur_id();
        }
        Ok(lhs)
    }

    fn parse_mul_term(&mut self) -> Result<Expr> {
        let mut start = self.cur_id();
        let mut lhs = self.parse_factor()?;
        while let Some(op_token) = accept!(self, TokenKind::Slash | TokenKind::Star) {
            // create binop node and push span
            let op = BinOp::new(&op_token.kind);
            self.ast_ctx
                .push_span(op.id, Some(op_token.id), Some(op_token.id));

            let rhs = self.parse_factor()?;
            lhs = Expr::bin_op(op, lhs, rhs);
            self.ast_ctx.push_span(lhs.id, start, self.prev_id());
            start = self.cur_id();
        }
        Ok(lhs)
    }

    fn parse_factor(&mut self) -> Result<Expr> {
        match self.peek() {
            Some(TokenKind::Int(_)) => self.parse_integer(),
            Some(TokenKind::Float(_)) => self.parse_float(),
            Some(TokenKind::Name(_)) => self.parse_variable(),
            Some(TokenKind::Minus) | Some(TokenKind::Plus) => self.parse_unary_factor(),
            Some(TokenKind::LParen) => self.parse_grouping(),
            Some(_) => unimplemented!(), //TODO: error handling
            None => Err(self.end_of_file_err()),
        }
    }

    fn parse_integer(&mut self) -> Result<Expr> {
        parse_node!(self, {
            let tok = expect!(self, TokenKind::Int(_))?;
            if let TokenKind::Int(ref n) = tok.kind {
                Expr::integer(n.clone())
            } else {
                unreachable!()
            }
        })
    }

    fn parse_float(&mut self) -> Result<Expr> {
        parse_node!(self, {
            let tok = expect!(self, TokenKind::Float(_))?;
            if let TokenKind::Float(ref n) = tok.kind {
                Expr::float(n.clone())
            } else {
                unreachable!()
            }
        })
    }

    fn parse_variable(&mut self) -> Result<Expr> {
        parse_node!(self, {
            let name = self.parse_var_name()?;
            Expr::variable(name)
        })
    }

    fn parse_unary_factor(&mut self) -> Result<Expr> {
        let start = self.cur_id();
        let op_token = expect!(self, TokenKind::Plus | TokenKind::Minus)?;

        // create binop node
        let op = UnaryOp::new(&op_token.kind);
        self.ast_ctx
            .push_span(op.id, Some(op_token.id), Some(op_token.id));

        let sub_expr = self.parse_factor()?;
        let expr = Expr::unary_op(op, sub_expr);
        self.ast_ctx.push_span(expr.id, start, self.prev_id());

        Ok(expr)
    }

    fn parse_grouping(&mut self) -> Result<Expr> {
        expect!(self, TokenKind::LParen)?;
        let expr = self.parse_expr()?;
        expect!(self, TokenKind::RParen)?;
        Ok(expr)
    }

    fn parse_var_name(&mut self) -> Result<VarName> {
        parse_node!(self, {
            let tok = expect!(self, TokenKind::Name(_))?;
            if let TokenKind::Name(ref n) = tok.kind {
                VarName::new(n.clone())
            } else {
                unreachable!()
            }
        })
    }

    /// Create an EOF error.
    ///
    /// For this, we create a span based on the end location of the last token.
    fn end_of_file_err(&self) -> ParserError {
        let token_id = self.tokens().last().map(|t| t.id);

        let span = if let Some(token_id) = token_id {
            self.input.token_ctx.span_from_id(token_id)
        } else {
            None
        };

        ParserError::SyntaxError(
            SyntaxError::UnexpectedEndOfFile,
            self.input.text_ctx.error_context(span),
        )
    }
}
