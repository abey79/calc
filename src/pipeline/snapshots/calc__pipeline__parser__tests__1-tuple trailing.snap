---
source: src/pipeline/parser.rs
expression: "parse(\"a = (1,);\").raw_ast"
---
Ast {
    stmts: [
        [1:1-9] Assign {
            name: [1:1-1] "a",
            value: [1:5-8] Tuple(
                [
                    [1:6-6] Integer(
                        1,
                    ),
                ],
            ),
        },
    ],
}
